Assist: mark individual nodes as finished

justinas opened the pull request:
Fixes https://github.com/gravitational/teleport.e/issues/1741 .

Makes sure that server ID is sent in the individual close messages. We do this in a quite naive way by sending the same `sessionData` that is sent when the session is established. This means that we possibly have some stale data (such as `last_active`), but for us only `server_id` is important. We could fix this by sending a leaner custom structure in this case.

Commit:
Mark nodes as done when command finishes in Assist

Commit:
Merge remote-tracking branch 'origin/master' into justinas/assist-mark-nodes-finished

Commit:
Split Close/CloseWithPayload

Commit:
Merge remote-tracking branch 'origin/master' into justinas/assist-mark-nodes-finished

Commit:
Expect a close message before the summary in test

Commit:
Remediate confusing type usage in test

`Envelope` is the outer layer used (in protobuf format)
for execution / terminal sessions.

Meanwhile `outEnvelope` is the inner layer
used (in JSON) spefically with assist (execution),
when outer envelope is of "raw" type

Using `Envelope` where decoding `outEnvelope` in tests
previously worked "by accident" due to matching field names

Commit:
Assert that server ID is set in the close message

jakule commented on web/packages/teleport/src/Assist/context/AssistContext.tsx:
```diff
@@ -465,30 +467,27 @@ export function AssistContextProvider(props: PropsWithChildren<unknown>) {
-              executeCommandWebSocket.current.send(bytearray.buffer);
-            }
-
+            executeCommandWebSocket.current.send(bytearray.buffer);
```
I think this line is no more needed. I forgot to remove it while I was working on this https://github.com/gravitational/teleport/pull/28456

jakule commented on lib/web/command.go:
```diff
@@ -674,7 +674,12 @@ func (t *commandHandler) streamOutput(ctx context.Context, tc *client.TeleportCl
 	}
 
-	if err := t.stream.SendCloseMessage(); err != nil {
+	sessionMetadataPayload, err := json.Marshal(siteSessionGenerateResponse{Session: t.sessionData})
```
I'm not sure if sending all session data is a good idea. We need only node ID, and session data may contain some stuff that we may not want to share (if not now it may in the future).

jakule committed:
Merge remote-tracking branch 'origin/master' into justinas/assist-mark-nodes-finished

jakule committed:
Refactor command execution logic and adjust WebSocket handling

This commit changes how command execution and WebSocket handling are performed in the code. Instead of manually managing session close signals and command execution notifications via WebSocket, we have wrapped it into more easily manageable form.

Changed parts: assist and websocket libraries, test and several components of 'teleport'.

Why:
This allows cleaner command management and error handling, leading to more reliable and maintainable application. Also made code more readable and easy to understand. Made WebSocket more precise with its handling to prevent cases where stale or incorrect data might interrupt the session.

Details:
Instead of sending a close signal after all commands have been executed, we now send individual session end updates for each command. Sessions no longer remain active unnecessarily.

Message handling in command execution has been refactored for better error propagation.

All these changes were also adjusted and reflected in the associated test cases.
Also, names and types of various data structures are edited to reflect their actual usage.

jakule commented:
@ryanclark Can you take a look at this?

ryanclark reviewed (approved):
Approving with @jakule changing `server_id` to `node_id` for consistency across message types

jakule committed:
Change ServerID to NodeID to improve code consistency

This commit replaces occurrences of `ServerID` with `NodeID` across several files.
`ServerID` was misleading and causing confusion as it was serving as an identifier for nodes, so to improve comprehension and consistency in the codebase, all instances of `ServerID` have been replaced with `NodeID`. Tests have also been updated to reflect this change.
